File: AtC/ABC/104/D.py
s = input()
ABC = [0 for c in s]
AB = [0 for c in s]
A = [0 for c in s]
qmarks=0
if s[0]=='A' or s[0]=='?': A[0]=1;

for i in range(0,len(s)):
    mult=1
    if (s[i]=='?'): mult=3
    A[i] = A[i-1]*mult
    AB[i] = AB[i-1]*mult
    ABC[i] = ABC[i-1]*mult

    if s[i]=='A':
        A[i] += + 3**qmarks
    elif s[i] == 'B':
        AB[i] += A[i-1]
    elif s[i] == 'C':
        ABC[i] += AB[i-1]
    else:
        A[i] += 3**qmarks
        AB[i] += A[i-1]
        ABC[i] += AB[i-1]
        qmarks+=1

print(ABC[-1]%(1e9+7))

File: AtC/ABC/104/D.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

ll m = 1e9+7;

int main() {
	string s; cin>>s;
	vector<ll> ABC(s.size()), AB(s.size()), A(s.size());
	ll qmarks=1;
	if (s[0]=='A' or s[0]=='?') A[0]=1;
	if (s[0]=='?') qmarks=3;

	for (ll i=1; i<s.size(); i++) {
		A[i] = A[i-1];
		AB[i] = AB[i-1];
		ABC[i] = ABC[i-1];
		if (s[i]=='A')
			A[i] = (A[i-1] + qmarks)%m;
		if (s[i]=='B')
			AB[i] = (AB[i-1] + A[i-1])%m;
		if (s[i]=='C')
			ABC[i] = (ABC[i-1] + AB[i-1])%m;
		if (s[i]=='?') {
			A[i] = ((A[i-1]*3)%m + qmarks)%m;
			AB[i] = ((AB[i-1]*3)%m + A[i-1])%m; 
			ABC[i] = ((ABC[i-1]*3)%m + AB[i-1])%m;
			qmarks = (qmarks*3)%m;
		}
	}
	
	cout << ABC.back() << endl;

	return 0;
}
